scene/3d/skeleton_3d.cpp

```diff
@@ -224,6 +224,61 @@ void Skeleton3D::_update_process_order() {
 	process_order_dirty = false;
 }
 
+Transform3D dual_quaternion_to_matrix(Quaternion Qn, Quaternion Qd)
+{
+	//Original version of this function is in https://users.cs.utah.edu/~ladislav/dq/dqs.cg
```

Comment:
You need to add the license note from the original code

Reply:
Hey so this is just a math reference. Do we think we need copyright? But actually going to that site I realize now it has a function we can use, the scaling which may fix my TODO item!

Reply:
Did you base your code on that? You say "original version" so that implies it's copied or based on it

Reply:
I actually based my code on someone else's GDscript code (https://godotengine.org/asset-library/asset/2062). The reference code is a cg shader not even the same language. But I am curious what we should do here. I have no problem the copyright notice would look something like this:

`Copyright (C) 2006-2007 University of Dublin, Trinity College, All Rights Reserved`

But really I wonder is it necessary since only a part of that code is "referenced" and the reference is just a math formula (convert dual quaternion to mat4).

This comment is actually from the GDScript code and I just carried it over.

Reply:
FYI its stated in the notice at the top of the link:
1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, **an acknowledgment in the product documentation would be
     appreciated but is not required.**

Reply:
That's for *using* the library though, for *copying* the source it says "This notice may not be removed or altered from any source distribution."

Reply:
I had a look, I can find a number of papers with similar formula, and the rotational part is also on Wikipedia: https://en.wikipedia.org/wiki/Dual_quaternion#Dual_quaternions_and_4%C3%974_homogeneous_transforms

Overall my layman understanding is that math formulas aren't copyrightable, since they're facts (math proofs could be copyrighted though). So I think we don't need to include the full license statement for borrowing this formula to convert a dual quaternion to a 4x3 matrix.

I would keep the link to the original source though, which is useful to check assumptions and double check that the code was correctly ported.